File: CPT104/pipe/src/simplex.ts
// Step 1: Create readable and writable streams
const { readable, writable } = new TransformStream();

// Step 2: Write data to the writable stream
async function writeData() {
  const message = "Hello, Pipe!";
  const encoder = new TextEncoder();
  const writer = writable.getWriter();
  await writer.write(encoder.encode(message));
  writer.releaseLock();
}

// Step 3: Read data from the readable stream
async function readData() {
  const decoder = new TextDecoder();
  const reader = readable.getReader();
  const { value, done } = await reader.read();
  const receivedMessage = decoder.decode(value);

  console.log("Received message:", receivedMessage);

  // Step 4: Close the readable stream
  if (!done) {
    await reader.cancel();
  }
}

await Promise.all([writeData(), readData()]);

File: CPT104/pipe/src/simplex.rs
use std::io::{Read, Write};
use std::os::unix::net::UnixStream;

pub fn main() -> std::io::Result<()> {
    // Step 1: Create a pipe
    let (mut read_end, mut write_end) = UnixStream::pair()?;

    // Step 2: Write data into the pipe
    let message = "Hello, Pipe!";
    write_end.write_all(message.as_bytes())?;

    // Step 3: Read data from the pipe
    let mut buffer = [0; 128];
    let bytes_read = read_end.read(&mut buffer)?;
    let received_message = String::from_utf8_lossy(&buffer[..bytes_read]);

    println!("Received message: {}", received_message);

    // No need for Step 4: Closing the pipe, as UnixStream will be closed automatically when they go out of scope

    Ok(())
}
